### PR TITLE
Greedyperm

### PR DESCRIPTION
**Pull request by ctralie**
I added a feature that allows users to intelligently subsample point clouds with furthest point sampling, and I added a notebook to show usage and appropriate unit tests.  It seems to be working well, but one of the tests on travis fails because it never gets past installing homebrew.

### CODE REVIEW

File: ripser/ripser.py [hunk 1]
```diff
@@ -37,6 +37,82 @@
 from pyRipser import doRipsFiltrationDMSparse as DRFDMSparse
 
 
+def get_greedy_perm_pc(X, N = -1, metric="euclidean"):
```
Comment (sauln) [inline ripser/ripser.py]: Could you run [Black](https://black.readthedocs.io/en/stable/) over all the code you touched please?

Comment (sauln) [inline ripser/ripser.py]: IDK what `pc` is when I first read this. Spelling it out completely would help.

Comment (sauln) [inline ripser/ripser.py]: I think `N=None` is a better convention for 'calculate a reasonable default'. This is debatable. Is there a reason for -1?

File: ripser/ripser.py [hunk 2]
```diff
@@ -37,6 +37,82 @@
...
+    Parameters
+    ----------
+    X: ndarray(N, d)
+        Point cloud with N points in d dimensions
+    N: int
+        Number of points to take in the permutation
+    metric: string or callable
+        The metric to use when calculating distance between instances in a 
+        feature array. If metric is a string, it must be one of the options 
+        specified in PAIRED_DISTANCES, including "euclidean", "manhattan",
```
Comment (sauln) [inline ripser/ripser.py]: what is `PAIRED_DISTANCES`?

File: ripser/ripser.py [hunk 3]
```diff
@@ -37,6 +37,82 @@
...
+        recorded. The callable should take two arrays from X as input and 
+        return a value indicating the distance between them.
+    Returns
+    -------
+    perm: ndarray(N)
+        Indices of points in the greedy permutation
+    lambdas: ndarray(N)
+        Covering radii at different points
+    """
+    if N == -1:
```
Comment (sauln) [inline ripser/ripser.py]: If `N=None`, then you could do `N = N or X.shape[0]`. Can we assume 0 is not a valid input?

File: ripser/ripser.py [hunk 4]
```diff
@@ -37,6 +37,82 @@
...
+        perm[i] = idx
+        lambdas[i-1] = ds[idx]
+        ds = np.minimum(ds, pairwise_distances(X, X[idx, :][None, :], metric=metric))
+    lambdas[-1] = np.max(ds)
+    return (perm, lambdas)
+
+
+
+def get_greedy_perm_dm(D, N = -1):
+    """
```
Comment (sauln) [inline ripser/ripser.py]: This function to the previous one except 2 lines. Can't we combine them into one?

File: ripser/ripser.py [hunk 5]
```diff
@@ -125,11 +225,29 @@ def ripser(
                 "The input point cloud has more columns than rows; "
                 + "did you mean to transpose?"
             )
-        X = pairwise_distances(X, metric=metric)
 
-    if not (X.shape[0] == X.shape[1]):
-        raise Exception("Distance matrix is not square")
-    dm = X
+    idx_perm = np.arange(X.shape[0])
```
Comment (sauln) [inline ripser/ripser.py]: There are a lot of forks right here. Can we put all the exception handling at the top all in one place? Then have this be just logic for getting dm?

What about an interface like:

```
if n_perm and n_perm > X.shape[0]:
    raise Exception
if n_perm and distance_matrix and sparse.issparse(X):
   raise Exception

if n_perm:
    dm, r_cover, idx_perm = _compute_permuted_distance_matrix(X, N=n_perm, distance_matrix=distance_matrix, metric=metric)
else:
    dm = X if distance_matrix else pairwise_distances(X, metric=metric)
    r_cover, idx_perm = 0.0, np.arange(X.shape[0])
```

File: test/test_ripser.py [hunk 6]
```diff
@@ -45,6 +45,18 @@ def test_non_square_dist_matrix(self):
 
         with pytest.raises(Exception):
             ripser(data, distance_matrix=True)
+    
+    def test_too_many_perm(self):
+        X = np.zeros((10, 2))
+
+        with pytest.raises(Exception):
```
Comment (sauln) [inline test/test_ripser.py]: you should be able to check the text in the exception is for the exception you expect as well.

File: test/test_ripser.py [hunk 7]
```diff
@@ -45,6 +45,18 @@ def test_non_square_dist_matrix(self):
 
         with pytest.raises(Exception):
             ripser(data, distance_matrix=True)
+    
+    def test_too_many_perm(self):
```
Comment (sauln) [inline test/test_ripser.py]: Could you also test with `n_perm=0` and `n_perm=None` and with default `n_perm`.
